### PR TITLE
search image in image performance refactoing

### PR DESCRIPTION
**Pull request by DanielWilczek**

### CODE REVIEW

File: G1ANT.Addon.Images/G1ANT.Addon.Images/Api/Imaging.cs [hunk 1]
```diff
@@ -46,97 +51,114 @@ public static bool AreImagesEqual(Bitmap image1, Bitmap image2)
...
-
-                    pBig += bigOffset;
                 }
+                y += 1;
             }
+        }
 
-            bigBmp.UnlockBits(bigData);
-            smallBmp.UnlockBits(smallData);
+        private static bool ContainSameElements(int[] first, int firstStart, int[] second, int secondStart, int length, double tolerance)
```
Comment (MarekKosinski) [inline G1ANT.Addon.Images/G1ANT.Addon.Images/Api/Imaging.cs]: You don't use the length argument

File: G1ANT.Addon.Images/G1ANT.Addon.Images/Api/Imaging.cs [hunk 2]
```diff
@@ -46,97 +51,114 @@ public static bool AreImagesEqual(Bitmap image1, Bitmap image2)
...
+        {
+            if (tolerance > 0)
```
Comment (MarekKosinski) [inline G1ANT.Addon.Images/G1ANT.Addon.Images/Api/Imaging.cs]: I think this whole function can be reduced to:

        private static bool ContainSameElements(int[] first, int firstStart, int[] second, int secondStart, int length, double tolerance)

        {
            return !second
                .Where((t, i) =>
                first[i + firstStart] - tolerance >= second[i + secondStart] ||
                first[i + firstStart] + tolerance <= second[i + secondStart])
                .Any();
        }

File: G1ANT.Addon.Images/G1ANT.Addon.Images/Api/Imaging.cs [hunk 3]
```diff
@@ -46,97 +51,114 @@ public static bool AreImagesEqual(Bitmap image1, Bitmap image2)
 
         public static Rectangle IsImageInImage(Bitmap smallBmp, Bitmap bigBmp, double tolerance)
```
Comment (MarekKosinski) [inline G1ANT.Addon.Images/G1ANT.Addon.Images/Api/Imaging.cs]: I know this is not your code, but on the 26 line there's unused, deprecated method, remove it for code clarity

File: G1ANT.Addon.Images/G1ANT.Addon.Images/Api/Imaging.cs [hunk 4]
```diff
@@ -46,97 +51,114 @@ public static bool AreImagesEqual(Bitmap image1, Bitmap image2)
...
+            }
 
-                bool matchFound = true;
+            var haystackArray = GetPixelArray(haystack);
+            var needleArray = GetPixelArray(needle);
+            var firstLineMatchPointsList = FindMatch(haystackArray.Take(haystack.Height - needle.Height), needleArray[0], margin);
 
-                for (int y = 0; y < bigHeight; y++)
+            var resultCollection = new ConcurrentBag<Point>();
+            ParallelLoopResult result = Parallel.ForEach<Point>(firstLineMatchPointsList, (firstLineMatchPoint) =>
```
Comment (MarekKosinski) [inline G1ANT.Addon.Images/G1ANT.Addon.Images/Api/Imaging.cs]: you are not using the result anywhere so it's no need to assigning it to a variable, also <Point> is redundant

File: G1ANT.Addon.Images/G1ANT.Addon.Images/Api/Imaging.cs [hunk 5]
```diff
@@ -46,97 +51,114 @@ public static bool AreImagesEqual(Bitmap image1, Bitmap image2)
...
+            });
 
-                                int inf = pBig[0] - margin;
-                                int sup = pBig[0] + margin;
-                                if (sup < pSmall[0] || inf > pSmall[0])
-                                {
-                                    matchFound = false;
-                                    break;
-                                }
+            if (resultCollection.Count() > 0)
```
Comment (MarekKosinski) [inline G1ANT.Addon.Images/G1ANT.Addon.Images/Api/Imaging.cs]: Use Any();

File: G1ANT.Addon.Images/G1ANT.Addon.Images/Api/Imaging.cs [hunk 6]
```diff
@@ -46,97 +51,114 @@ public static bool AreImagesEqual(Bitmap image1, Bitmap image2)
...
-            int smallWidth = smallBmp.Width * 3;
-            int smallHeight = smallBmp.Height;
+            Rectangle rectangleResult = FindImage(smallBmp, bigBmp, tolerance);
+            if (rectangleResult != Rectangle.Empty) { return rectangleResult; }
+            else { return Rectangle.Empty; }
+        }
 
+        private static Rectangle FindImage(Bitmap needle, Bitmap haystack, double tolerance)
+        {
             Rectangle location = Rectangle.Empty;
```
Comment (MarekKosinski) [inline G1ANT.Addon.Images/G1ANT.Addon.Images/Api/Imaging.cs]: Use var. Also you are assigning value to a variable which you aren't really changing during the method execution. It's pretty useless

File: G1ANT.Addon.Images/G1ANT.Addon.Images/Api/Imaging.cs [hunk 7]
```diff
@@ -46,97 +51,114 @@ public static bool AreImagesEqual(Bitmap image1, Bitmap image2)
...
             int margin = Convert.ToInt32(255.0 * tolerance);
 
-            unsafe
+            if (null == haystack || null == needle)
```
Comment (MarekKosinski) [inline G1ANT.Addon.Images/G1ANT.Addon.Images/Api/Imaging.cs]: These conditions can be merged with conditions on 68 lines and extracted to a one-liner method for clarity

File: G1ANT.Addon.Images/G1ANT.Addon.Images/Api/Imaging.cs [hunk 8]
```diff
@@ -46,97 +51,114 @@ public static bool AreImagesEqual(Bitmap image1, Bitmap image2)
...
-                                    matchFound = false;
-                                    break;
-                                }
+            if (resultCollection.Count() > 0)
+            {
+                var rectangle = resultCollection.FirstOrDefault();
+                return new Rectangle(rectangle.X, rectangle.Y, needle.Width, needle.Height);
+            }
+            else return Rectangle.Empty;
+        }
```
Comment (MarekKosinski) [inline G1ANT.Addon.Images/G1ANT.Addon.Images/Api/Imaging.cs]: That's how I would see this method:

```
private static Rectangle FindImage(Bitmap needle, Bitmap haystack, double tolerance)
        {
            var location = Rectangle.Empty;
            var margin = Convert.ToInt32(255.0 * tolerance);
            if (IsCorrectNeedleAndHaystackProperties(needle, haystack))
            {
                var haystackArray = GetPixelArray(haystack);
                var needleArray = GetPixelArray(needle);
                var firstLineMatchPointsList = FindMatch(haystackArray.Take(haystack.Height - needle.Height), needleArray[0], margin);
                var resultCollection = new ConcurrentBag<Point>();

                Parallel.ForEach(firstLineMatchPointsList, firstLineMatchPoint =>
                {
                    if (IsNeedlePresentAtLocation(haystackArray, needleArray, firstLineMatchPoint, 1, margin))
                    {
                        resultCollection.Add(firstLineMatchPoint);
                    }
                });

                if (resultCollection.Any())
                {
                    var rectangle = resultCollection.FirstOrDefault();
                    location = new Rectangle(rectangle.X, rectangle.Y, needle.Width, needle.Height);
                }
            }

            return location;
        }

        private static bool IsCorrectNeedleAndHaystackProperties(Bitmap needle, Bitmap haystack)
        {
            return haystack != null && needle != null && haystack.Width >= needle.Width && haystack.Height >= needle.Height;
        }
```

File: G1ANT.Addon.Images/G1ANT.Addon.Images/Api/Imaging.cs (hunk 2 above)
Comment (DanielWilczek) [reply G1ANT.Addon.Images/G1ANT.Addon.Images/Api/Imaging.cs]: tolerance is not included into logic

Comment (MarekKosinski) [reply G1ANT.Addon.Images/G1ANT.Addon.Images/Api/Imaging.cs]: It is included, but as we talked, your solution is faster

**Pull request merged**